### PR TITLE
[CARBONDATA-530] Keep the CarbonDecoder on top of limit in case of limit+sort plan

### PR DESCRIPTION
**Pull request by ravipesala**
There is a new optimization in spark2.1, it can group limit and sort together and get the topN efficiently. so we should put outer decoder above limit to make use of this optimization.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1199/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1210/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1211/

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonDictionaryDecoder.scala [hunk 1]
```diff
@@ -371,6 +307,79 @@ case class CarbonDictionaryDecoder(
 
 }
 
+object CarbonDictionaryDecoder {
+
+  def convertOutput(output: Seq[Attribute],
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/CarbonDictionaryDecoder.scala]: please give some description comment for this function

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonDictionaryDecoder.scala [hunk 2]
```diff
@@ -371,6 +307,79 @@ case class CarbonDictionaryDecoder(
...
+        } else {
+          a
+        }
+      } else {
+        a
+      }
+    }
+  }
+
+  def canBeDecoded(attr: Attribute, profile: CarbonProfile): Boolean = {
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/CarbonDictionaryDecoder.scala]: please give some description comment for this function

**Comment by jackylk** [comment]
LGTM except for adding comments

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1214/

**Pull request closed without merging**